runtime/lua/vim/snippet.lua

```diff
@@ -0,0 +1,66 @@
+local grammar = require('_snippet_grammar')
+
+--- Returns the current cursor position (0-based).
+--- @return { [1]: integer, [2]: integer }
+local function cursor_position()
+  local cursor = vim.api.nvim_win_get_cursor(0)
+  return { cursor[1] - 1, cursor[2] }
+end
+
+--- Resolves variables.
+--- @param var string
+--- @return integer|string|nil
+local function get_variable_value(var)
+  if var == 'TM_SELECTED_TEXT' then
+    return ''
```

Comment:
LSP defines this variable as the currently selected text or the empty string. However snippet expansion/insertion will always (?) happen in insert mode, so there won't be a visual selection, and I don't want to use `'<`/`'>` since it might not be intended for the last visual selection to be inserted here.

Reply:
If snippet expansion always happens in insert-mode, what is the intention of this case?

Note that vim has `:help Select-mode` (as opposed to `:help Visual-mode`) which behaves more like selections in other text editors. Unfortunately It sets `'<`/`'>`, and `visualmode()` doesn't give a hint about whether the last mode was select-mode or visual-mode. But `mode()` does the right thing if we are currently in select-mode.

Reply:
The issue here is that the LSP spec describes (poorly) how variables should work with defaults, even if that combination doesn't make sense in the vast majority of cases; this is one of the "nonsense" combinations. Nevertheless, if we want to implement the spec, we need to handle them somehow.

Reply:
(This case is much more relevant for snippet _plugins_ that implement manual triggers; it's not relevant to expanding LSP snippets from completion items.)

Reply:
Right. I also need to have this case because `TM_SELECTED_TEXT` is a defined variable, and so even if we noop for it we would insert its default if present.